Update docs for GDScript 2.0

vnen opened the pull request:
GDScript syntax will change a bit after https://github.com/godotengine/godot/pull/39093, so this is an update of the basics and export pages to fit the new way.

I removed the section for exporting arrays because it's still not supported, but I'll add it back once typing is re-added and there is a syntax for typed arrays.

I also removed the groups/categories sections because I plan to add those as annotations.

vnen committed:
Update GDScript basics doc to follow GDScript 2.0

vnen committed:
Update export docs to use annotations

mhilbrunner commented:
Looks good so far, will have a more in-depth look later. Should we mark this as draft until https://github.com/godotengine/godot/pull/39093 is finished/merged?

vnen commented:
> Should we mark this as draft until [godotengine/godot#39093](https://github.com/godotengine/godot/pull/39093) is finished/merged?

Sure, if you prefer.

vnen committed:
Describe GDScript properties

vnen commented:
Now that https://github.com/godotengine/godot/pull/40598 is merged we can also merge this.

Calinou commented on getting_started/scripting/gdscript/gdscript_basics.rst:
```diff
@@ -29,48 +29,48 @@ here's a simple example of how GDScript looks.
 
-    # (optional) class definition with a custom icon
+    # (optional) class definition:
+    class_name myclass
```
Shouldn't it use PascalCase?

```suggestion
    class_name MyClass
```

Calinou commented on getting_started/scripting/gdscript/gdscript_basics.rst:
```diff
@@ -83,25 +83,28 @@ here's a simple example of how GDScript looks.
+        var local_var = param1 + 3
+        return local_var
 
+    # Functions override functions with the same name on the base/super class.
```
Does this apply to built-in methods such as `_process()`? In 3.2, these are called automatically.

vnen replied:
Actually, I still need to disable the multilevel call in the code, but that is in the plan. I didn't do it on the main PR because it touches a bunch of code outside GDScript so it should be reviewed independently.

Calinou commented on getting_started/scripting/gdscript/gdscript_basics.rst:
```diff
@@ -294,11 +283,102 @@ Literals
++------------------------------+---------------------------------------------------------------------------------------------------+
+| ``@tool``                    | Enable the `Tool mode`_.                                                                          |
++------------------------------+---------------------------------------------------------------------------------------------------+
+| ``@onready``                 | Defer initialization of variable until the node is in the tree, See `Onready annotation`_).       |
```
```suggestion
| ``@onready``                 | Defer initialization of variable until the node is in the tree. See `Onready annotation`_.        |
```

Calinou commented on getting_started/scripting/gdscript/gdscript_basics.rst:
```diff
@@ -294,11 +283,102 @@ Literals
++------------------------------+---------------------------------------------------------------------------------------------------+
+| ``@onready``                 | Defer initialization of variable until the node is in the tree, See `Onready annotation`_).       |
++------------------------------+---------------------------------------------------------------------------------------------------+
+| ``@icon(path)``              | Set the class icon to show in editor. To be used together with the ``class_name`` keyword         |
```
```suggestion
| ``@icon(path)``              | Set the class icon to show in editor. To be used together with the ``class_name`` keyword.        |
```

Calinou commented on getting_started/scripting/gdscript/gdscript_basics.rst:
```diff
@@ -383,8 +463,24 @@ Strings can contain the following escape sequences:
+:ref:`StringName <class_StringName>`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Immutable strings that allows only one instance of each name. They are slower to
```
```suggestion
An immutable string that allows only one instance of each name. They are slower to
```

Calinou commented on getting_started/scripting/gdscript/gdscript_basics.rst:
```diff
@@ -383,8 +463,24 @@ Strings can contain the following escape sequences:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Immutable strings that allows only one instance of each name. They are slower to
+create and may result in waiting for locks when multithreading. But they are
+very fast to compare which makes good candidates for dictionary keys.
```
```suggestion
create and may result in waiting for locks when multithreading. In exchange, they're
very fast to compare, which makes them good candidates for dictionary keys.
```

Calinou commented on getting_started/scripting/gdscript/gdscript_basics.rst:
```diff
@@ -394,6 +490,12 @@ Vector built-in types
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Same as a Vector2 but the components are integers. Useful for representing
+items in a grid.
```
For consistency with the Vector3i description:

```suggestion
items in a 2D grid.
```

vnen committed:
Apply suggestions from code review

Include suggestions from Calinou

Co-authored-by: Hugo Locurcio <hugo.locurcio@hugo.pro>

YeldhamDev commented on getting_started/scripting/gdscript/gdscript_basics.rst:
```diff
@@ -83,25 +83,28 @@ here's a simple example of how GDScript looks.
-        var local_var2 = param1 + 3
-        return local_var2
-
+        match param3:
+            3: print("param3 is 3!")
```
This should be indented like the one below.

NathanLovato replied:
```suggestion
            3:
                print("param3 is 3!")
```

YeldhamDev commented on getting_started/scripting/gdscript/gdscript_basics.rst:
```diff
@@ -756,22 +888,29 @@ return early with the ``return`` keyword, but they can't return any value.
+        var result = []
+        for item in arr:
+            result.push_back(function.call(item))
+        return result
 
-    # Store a function reference.
-    var my_func = funcref(my_node, "my_function")
-    # Call stored function reference.
-    my_func.call_func(args)
+    func add1(value: int) -> int:
+        return value + 1;
 
+    func _ready() -> void:
+        var my_array = [1, 2, 3]
+        var plus_one = map(my_array, add1)
+        print(plus_one) # Prints [2, 3, 4].
```
There should be 2 lines of space between those methods.

vnen replied:
Using two lines for separation on small doc snippets isn't helpful. I understand the style guide says so, but in small snippets it just looks off. That's why I asked an exception on documentation pages: https://github.com/godotengine/godot-docs/issues/3613#issuecomment-636367359

Xrayez replied:
Yes, the style *guide* shouldn't *mandate* style.

YeldhamDev commented on getting_started/scripting/gdscript/gdscript_basics.rst:
```diff
@@ -1104,43 +1275,43 @@ the ``is`` keyword can be used::
+If you need to call a different function from the super class, you can specify
+the function name with the attribute operator::
+
+    func overriding():
+        return 0 # This overrides the method in the base class.
+
+    func dont_override():
+        return super.overriding() # This calls the method as defined in the base class.
```
2 lines of space here too.

YeldhamDev commented on getting_started/scripting/gdscript/gdscript_basics.rst:
```diff
@@ -1481,137 +1648,69 @@ Our ``BattleLog`` node receives each element in the binds array as an extra argu
+function won't give the control back to the caller::
 
-        return result
+    func no_wait():
+        var x = await get_five()
+        print("This doesn't make this function a coroutine.")
 
+    func get_five():
+        return 5
```
2 lines of space.

YeldhamDev commented on getting_started/scripting/gdscript/gdscript_basics.rst:
```diff
@@ -1481,137 +1648,69 @@ Our ``BattleLog`` node receives each element in the binds array as an extra argu
-        if result is GDScriptFunctionState: # Still working.
-            result = yield(result, "completed")
+    func get_signal():
+        return $Button.button_up
 
-        return result
+    func wait_button():
+        await get_signal()
+        print("Button was pressed")
```
2 lines of space.

Xrayez commented:
> I also removed the groups/categories sections because I plan to add those as annotations.

I've added those sections in #3444. Can annotations provide a way to add groups and categories dynamically? Perhaps this can be adapted to "Adding script groups and categories dynamically via code". Maybe it deserves more detailed documentation on `_get_propety_list()` (not necessarily GDScript documentation, but C++). Unless we're moving away from using `_get_propety_list()` in GDScript at all... 😮

See example use case in https://github.com/godotengine/godot/pull/10303#issuecomment-618979118.

vnen commented:
> > I also removed the groups/categories sections because I plan to add those as annotations.
> 
> I've added those sections in #3444. Can annotations provide a way to add groups and categories dynamically? Perhaps this can be adapted to "Adding script groups and categories dynamically via code". Maybe it deserves more detailed documentation on `_get_propety_list()` (not necessarily GDScript documentation, but C++). Unless we're moving away from using `_get_propety_list()` in GDScript at all... 
> 
> See example use case in [godotengine/godot#10303 (comment)](https://github.com/godotengine/godot/pull/10303#issuecomment-618979118).

Annotations cannot do anything dynamically, they are only evaluated at compile-time.

I guess we can leave this info here but should be worded as an advanced method for dynamic changes (I know right now it's the only way to do it, but annotations for this should come before Godot 4 is released). So if you want I can remove the deletion.

Though I wish this could be approved and merged soon, as it already have conflicts again.

Xrayez commented:
I think that's fine to remove those sections for now, I was just curious how it plays with existing annotations system. As I said, it probably needs a better introduction on `_get_property_list()` in the first place.

NathanLovato committed:
Fix two code lines wrapped in a single one

NathanLovato commented:
Hey @vnen it's looking good. I just fixed the remaining review comment. If you can fix the conflicts, I'll merge your PR.

artism90 commented on getting_started/scripting/gdscript/gdscript_basics.rst:
```diff
@@ -1104,43 +1276,43 @@ the ``is`` keyword can be used::
-To call a function in a *parent class* (i.e. one ``extend``-ed in your current
-class), prepend ``.`` to the function name::
+To call a function in a *super class* (i.e. one ``extend``-ed in your current
+class), user the ``super`` keyword::
```
```suggestion
class), use the ``super`` keyword::
```

akien-mga commented on getting_started/scripting/gdscript/gdscript_exports.rst:
```diff
@@ -37,99 +39,102 @@ Examples
     # Editor will enumerate as THING_1, THING_2, ANOTHER_THING.
     enum NamedEnum {THING_1, THING_2, ANOTHER_THING = -1}
-    export(NamedEnum) var x
+    @export var x: NamedEnum
```
This doesn't seem to work for me in the current `master` branch (godotengine/godot@a6bb4f70e07edc52cb969548876a0b57abeef546).

vnen replied:
Yeah, I need to fix the export because it's missing some stuff still.

NathanLovato commented:
@vnen I'd like to merge this and other GDScript-related PRs, even if there'll be more changes in the coming weeks or months, to avoid merge conflicts down the road. Can I go ahead and rebase and merge?

vnen commented:
@NathanLovato I'll try to take sometime this week to fix the conflicts. Unless someone else is willing to do it sooner.

NathanLovato commented:
I can do it. Unless you want to make changes, you can leave it to me.

NathanLovato commented:
I had to open a new PR to fix the merge conflicts due to the rebase on master: #4247. Closing this one and merging the new PR
